src/lib/Inference.hs

```diff
@@ -2261,41 +2247,32 @@ getSuperclassClosure projs givens newGivens = do
       ty <- getType x
       modify \(Givens m) -> Givens $ HM.insert (EKey ty) x m
 
-tryApplyM :: (MonadFail1 m, Builder m) => Atom n -> Given n -> m n (Given n)
-tryApplyM proj dict = do
-  dictTy <- getType dict
-  projTy <- getType proj
-  instantiateProjParams projTy dictTy >>= \case
-    NothingE -> fail "couldn't instantiate params"
-    JustE (ListE params) -> liftBuilder $ buildBlock do
-      Distinct <- getDistinct
-      instantiated <- naryApp (sink proj) (map sink params)
+getDirectSuperclasses :: EnvReader m => Given n -> m n [Given n]
+getDirectSuperclasses dict = do
+  numSuperclasses <- getNumSuperclasses dict
+  forM [0..numSuperclasses-1] \i -> do
+    liftBuilder $ buildBlock do
       dictAtom <- emitBlock $ sink dict
-      app instantiated dictAtom
-    _ -> error "impossible"
-
-instantiateProjParams :: EnvReader m => Type n -> Type n
-                      -> m n (MaybeE (ListE Atom) n)
-instantiateProjParams projTy dictTy = do
-  PairE projTy' dictTy' <- sinkM $ PairE projTy dictTy
-  result <- liftSolverM $ solveLocal do
-    params <- instantiateProjParamsM (sink projTy') (sink dictTy')
-    zonk $ ListE params
-  case result of
-    Success params -> return $ JustE $ params
-    Failure _ -> return NothingE
-
-instantiateProjParamsM :: (EmitsInf n, Unifier m) => Type n -> Type n -> m n [Atom n]
-instantiateProjParamsM projTy dictTy = case projTy of
-  Pi (PiType (PiBinder b argTy ImplicitArrow) _ resultTy) -> do
-    param <- freshInferenceName argTy
-    resultTy' <- applyAbs (Abs b resultTy) param
-    params <- instantiateProjParamsM resultTy' dictTy
-    return $ Var param : params
-  Pi (PiType (PiBinder _ argTy PlainArrow) _ _) -> do
-    unify dictTy argTy
-    return []
-  _ -> error $ "unexpected projection type: " ++ pprint projTy
+      return $ getProjection [i] $ getProjection [0, 0] dictAtom
+
+  -- TypeCon sourceName dataDefName _ <- getType dict
+getNumSuperclasses :: EnvReader m => Given n -> m n Int
```

Comment:
Perhaps instead of counting only to iterate later, you could return the list of superclass types that you then could iterate over.

Reply:
yeah, that's better